accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -2748,20 +2748,30 @@ To match a pattern `p` against a value `v`:
         1.  Evaluate the key `expression` to `k` and call `containsKey(k)` on
             the value. If this returns `false`, the map does not match.
 
-        2.  Otherwise, evaluate `v[k]` and match the resulting value against
-            this entry's value subpattern. If it does not match, the map does
-            not match.
+        2.  Otherwise, evaluate `v[k]` to `r`.
+
+        3.  If `V` is non-nullable and `r` is `null` then the map does not
+            match.
+
+            *The `[]` operator on `Map` is always nullable. If we don't
+            explicitly check for `null` here, then a bad `Map` implementation
+            that returns `true` from `containsKey()` and then `null` from `v[k]`
+            for the same key would lead to a soundness hole when `V` is a
+            non-nullable type.*
+
+        4.  Else, match `r` against this entry's value subpattern. If it does
+            not match, the map does not match.
 
         A compiler is free to call `v[k]` and `containsKey()` in either order,
         or to elide calling one or both if it determines that doing so will
         produce the same result. It may assume that the map adheres to the
         following protocol:
 
-        *   If `containsKey(k)` returns `false` for some key, then `v[k]` will
+        *   When `containsKey(k)` returns `false` for some key, then `v[k]` will
             return `null`.
 
-        *   If `containsKey(k)` returns `true` for some key, then `v[k]` returns
-            an instance of the map's value type.
+        *   When `containsKey(k)` returns `true` for some key, then `v[k]`
```

Comment:
I think we're talking about getting rid of this, but note that this is incoherent with the above: if the compiler can assume this, it is free to elide the checks that you are specifying above.

Reply:
Yeah, I got this wrong. Fixed.